### PR TITLE
Feature/support json patch

### PR DESCRIPTION
**Pull request by mvcatsifma**
Support JSON Merge Patch

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -23,3 +25,10 @@ gulp.task('watch', function () {
     gulp.start('default', done);
   }));
 });
+
+gulp.task('jshint', function () {
+  return gulp.src(['src/*.js'])
+    .pipe(jshint('./.jshintrc'))
+    .pipe(jshint.reporter('jshint-stylish'))
+    .pipe(jshint.reporter('fail'));
+});
```
Comment (jcassee) [inline gulpfile.js:33]: If `jshint` is quick we could add this to the default task.

File: gulpfile.js [hunk 2]
```diff
@@ -4,6 +4,8 @@ var gulp = require('gulp'),
   batch  = require('gulp-batch'),
   concat = require('gulp-concat'),
   ignore = require('gulp-ignore'),
+  gulpIf = require('gulp-if'),
```
Comment (jcassee) [inline gulpfile.js]: What is `gulpIf`used for?

File: src/resource.js [hunk 3]
```diff
@@ -309,6 +309,31 @@ angular.module('hypermedia')
...
+       *
+       * @function
+       * @returns {object}
+       */
+      $patchRequest: {value: function (data) {
+        return {
+          method: 'patch',
+          url: this.$uri,
+          data: data,
+          headers: {'Content-Type': 'application/json'}
```
Comment (jcassee) [inline src/resource.js]: This should probably be `application/merge-patch+json`.

File: src/resource.js [hunk 4]
```diff
@@ -397,6 +422,27 @@ angular.module('hypermedia')
...
+      $merge: {value: function (data){
+        var self = this;
+        Object.keys(data).forEach(function(key){
+          if (data[key] === null) {
+            delete self[key];
+          } else {
+            self[key] = data[key];
+          }
+          return self;
+        });
```
Comment (jcassee) [inline src/resource.js]: This algorithm does not do a deep merge of objects [as described in the RFC](https://tools.ietf.org/html/rfc7386#section-2). It probably should.

**Comment by jcassee** [comment]
One more thing that is missing is an update of the README to document the `$patch` method. I can also write that if you want.

File: gulpfile.js (hunk 1 above)
Comment (mvcatsifma) [inline gulpfile.js:33]: Done.

File: gulpfile.js (hunk 2 above)
Comment (mvcatsifma) [inline gulpfile.js]: Not used, removed.

File: src/resource.js (hunk 3 above)
Comment (mvcatsifma) [inline src/resource.js]: Fixed.

File: src/resource.js (hunk 4 above)
Comment (mvcatsifma) [inline src/resource.js]: Fixed. Thanks :)

File: dist/hypermedia.js [hunk 5]
```diff
@@ -198,6 +198,28 @@ angular.module('hypermedia')
...
+       * @param {Resource} resource
+       * @returns a promise that is resolved to the resource
+       * @see Resource#$patchRequest
+       */
+      httpPatch: {value: function (resource, data) {
+        var self = this;
+        busyRequests += 1;
+        var request = updateHttp(resource.$patchRequest(data));
+        return $http(request).then(function () {
+          Resource.prototype.$merge.call(resource, request.data);
```
Comment (jcassee) [inline dist/hypermedia.js]: Good point using `request.data` here and not `data` from the arguments.

File: src/context.js [hunk 6]
```diff
@@ -111,6 +111,28 @@ angular.module('hypermedia')
...
+       * @param {Resource} resource
+       * @returns a promise that is resolved to the resource
+       * @see Resource#$patchRequest
+       */
+      httpPatch: {value: function (resource, data) {
+        var self = this;
+        busyRequests += 1;
+        var request = updateHttp(resource.$patchRequest(data));
+        return $http(request).then(function () {
+          Resource.prototype.$merge.call(resource, request.data);
```
Comment (jcassee) [inline src/context.js:126]: Good point using `request.data` here and not `data` from the arguments, by the way!

File: package.json [hunk 7]
```diff
@@ -10,12 +10,16 @@
     "gulp": "^3.9",
     "gulp-batch": "^1.0",
     "gulp-concat": "^2.6",
+    "gulp-if": "^2.0.0",
```
Comment (jcassee) [inline package.json:13]: This can go too, now. I'll remove it.

**Pull request merged**